### DISCUSSION TITLE
Why are function attributes not allowed?

### DISCUSSION THREAD

**Discussion by dimaqq** [microsoft/pyright#8838] [General]
MRE

```py
  def foo(a: int): return a

E foo.bar = 42 # E: Cannot assign to attribute "bar" for class "function"   Attribute "bar" is unknown
```

How did I get here, simplified:

```py
def deco(f: Callable[[], int]) -> Callable[[], int]:
    def inner():
        if inner.bar:
            print("returning cached value")
            return inner.bar["cache"]

        inner.bar["cache"] = f()
        return inner.bar["cache"]

    inner.bar: Dict[Literal["cache"], int] = {}  # cache
    return inner
```

Code tested on py3.8 .. 3.13

P.S. I'm not too happy with the (ilelgal?) type hint `inner.bar: SomeType = ...`, but Python allows this at run time.

**Comment by erictraut** [comment]
This isn't allowed for the same reason that setting an undeclared attribute isn't allowed on any other object. 

For example, this is allowed at runtime, but it's a type violation.

```python
class Foo: ...
x = Foo()
x.bar = 42  # Type error
```

If you want pyright to allow accesses to undeclared attributes on a function, you can disable the `reportFunctionMemberAccess` check.

If you want to do this in a type-safe manner, you would need to define a wrapper class that declares the attributes and their types. Something like this:

```python
class CallableWithBar[**P, R]:
    bar: dict[str, int] = {}

    def __init__(self, func: Callable[P, R]) -> None:
        self._func = func

    def __call__(self, *args: P.args, **kwargs: P.kwargs) -> R:
        return self._func(*args, **kwargs)
```

**Comment by dimaqq** [comment]
So far my crude work-around is a Protocol and one `# type: ignore`:

```py
def foo(a: int): return a

class Foo(Protocol):
    def __call__(self, a: int) -> int: ...
    bar: int

foo: Foo = foo  # type: ignore
```

After which pyright is happy with attribute access... and even mypy can be coerced (must have separate names for function def and typed variable).

P.S. I think what I wanted, ultimately, was to have correct types **without runtime overhead**.

**Comment by STofone** [comment]
If you use decorator to add function attributes, this may help. 
```python
from typing import Protocol, cast


class Bar(Protocol):
    def __call__(self): ...

    x: bool


def add_x(func):
    func.x = True
    return cast(Bar,func)


@add_x
def bar():
    print(bar.x)


bar()
```

**Reply by dimaqq** [reply]
This is a cool trick, thank you!
